### ISSUE TITLE
how to implement vector of high-order derivatives?

### ISSUE THREAD

**Issue by ziofil** [google/jax#1678]
I need to implement a vector of high-order derivatives `[f, D(f), D(D(f)), D(D(D(f))), ...]`, is there a recommended way to do this?
After defining the functions that it contains, I wonder if there is a way to evaluate it with a single call instead of multiple ones.

**Comment by levskaya** [comment]
Are you looking for something like this?

```python
import jax
from jax import numpy as np
from matplotlib import pyplot as plt

def make_derivs_vec(f, n=3):
  Ds = [f,]
  for i in range(n):
    Ds.append(jax.grad(Ds[-1]))
  return jax.jit(lambda x: np.array([d(x) for d in Ds]))

example = make_derivs_vec(lambda x: np.tanh(x * x), 3)
plt.plot([example(x) for x in np.arange(-np.pi, np.pi, 0.01)]);
```

this returns a function that evaluates all the derivatives in a single call.

**Comment by ziofil** [comment]
Yes! Thank you. I had also constructed the list by appending the gradient of the last item, but i had not thought about using jit on the lambda expression like you did.

**Comment by mattjj** [comment]
There's more you can do here, leveraging the [Faa di Bruno formula](https://en.wikipedia.org/wiki/Fa%C3%A0_di_Bruno%27s_formula) and also tricks for specific primitives. @jessebett is working on a JAX implementation of those things (in the jet branch and also #1185, likely much more on his fork), though it's not ready yet IIUC.

@levskaya 's solution works, but there's no work sharing between the evaluations (unless XLA finds all the common subexpressions for us, which it might). I suspect with the right nesting of `jvp`'s you can do work sharing at the autodiff level. Maybe see [this gist](https://gist.github.com/mattjj/02ae95e51ead485223245d343efbd461) for inspiration.
